### ISSUE TITLE
display equations on lines

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/graphing-quadratics#21]
Display equations on all lines (curved and straight) on the graph.   Ideally we'd like to minimize overlap of things on the graph, but that's a complicated problem.  I'm going to start with the same approach that was used in Graphing Lines.

**Comment by pixelzoom** [comment]
In graphing-lines, the equation is created in LineNode's constructor at line 95, based on whether `options.equationType` was specified (`null` by default). The position on the equation is set in LineNode `update`, in the `if` statement at line 209.

**Comment by pixelzoom** [comment]
For _Explore_ and _Standard Form_ screens, the equations need to be in standard form.  In _Vertex Form_ and _Focus & Directrix_ screens, they need to be in vertex form.   Coefficients to be shown in decimal format.

**Comment by pixelzoom** [comment]
Equations are placed on lines in [1.0.0-dev.14](https://phet-dev.colorado.edu/html/graphing-quadratics/1.0.0-dev.14/phet/graphing-quadratics_en_phet.html).  

The current algorithm is:

- For straight lines: position the equation above the left end of line, in graph quadrants 2 or 3.

- For parabolas: If vertex.x >= 0, position the equation to the left of the parabola. Otherwise position the equation to the right of the parabola. (So pick the side where the most space is available.) Place the equation at the edge of the graph, parallel to the tangent at the point where the parabola meets the edge of the graph. (So try to keep it towards the edge of the graph, where it's least likely to overlap vertex, focus, etc.)

@amanda-phet Over to you for review.  FYI, this was incredibly complicated to implement. And there are definitely some case where things overlap (e.g. _Vertex Form_ screen with a=1, h=0, k=1). Eliminating all overlap is neither cost effective or feasible, so please prioritize change requests and be prepared to compromise.

**Comment by pixelzoom** [comment]
Algorithms of a couple of other lines:

- For directrix: If vertex.x >= 0, position the equation at the left end of the line, otherwise at right end of line.  Position the equation above the line when directrix <= 9, below the line when directrix > 9.

- For axis of symmetry:  If vertex.y >= 0, position the equation at the bottom end of the line, otherwise at the top end of the line.  If vertex.x >= 0, position the equation to the right of the line, otherwise to the left of the line (to avoid overlapping the y axis).

**Comment by pixelzoom** [comment]
There are also situations where parabola equations go slightly off the edge of the graph.  That shouldn't be difficult to fix, but I'll wait until I hear whether placing parabola equations towards the edges of the graph is an acceptable strategy.

**Comment by amanda-phet** [comment]
I think overall the algorithms make sense. I've found instances of the equation running off the graph and agree there are some overlaps that are undesirable, but I'm ok with them being concessions. I think it's important @ariel-phet and @kathy-phet play with this quite a bit, too. 

I have played with this for quite a while this morning, and on first glance there were things I wanted to change, but after playing further I think these rules actually make the most sense.

**Comment by pixelzoom** [comment]
Thanks @amanda-phet. I'll proceed with fixing the cases where the equation runs off the graph. Then we can turn it over to the team to play with.

**Comment by pixelzoom** [comment]
In version [1.0.0-dev.17](https://phet-dev.colorado.edu/html/graphing-quadratics/1.0.0-dev.17/phet/graphing-quadratics_en_phet.html) the equations now remain fully on the graph. 

The only serious overlapping occurs in the _Vertex Form_ screen, when |k| > 6.  In this case, the vertex coordinates overlap the equation (see screenshot).  If you uncheck "Coordinates", then the overlap is not an issue (see screenshot).  But if we combine visibility of coordinates and equations into one checkbox (see #35) then this would hide the equation also. 

Coordinate checkbox checked:
![screenshot_766](https://user-images.githubusercontent.com/3046552/46239806-106fc200-c35b-11e8-8e18-f4dcd835b133.png)

Coordinate checkbox unchecked:
![screenshot_767](https://user-images.githubusercontent.com/3046552/46239807-14034900-c35b-11e8-92b5-a7963513d425.png)
 
I tried locating the unrotated equation to the left/right of the parabola in these cases, and it worked well for this screen, but caused problems in the _Focus & Directrix_ screen.  And it's difficult to make this behave differently per screen.  @amanda-phet do you want to do anything about this?

**Comment by amanda-phet** [comment]
This is great! And addresses the one concern I had for #35 (which I see now I never submitted). 

My thought was previously: Given that there are instances of overlap between vertex and equation, or equation and equation, I think we *do* need a control to show/hide equations. 
<img width="822" alt="screen shot 2018-09-27 at 10 44 35 am" src="https://user-images.githubusercontent.com/7852916/46164449-3c9c1d80-c24b-11e8-9981-3dc33b782ce6.png">

But with what you've done to address that scenario here, I think we don't necessarily need that control. I will respond directly in that issue.

All is looking great to me!

**Comment by pixelzoom** [comment]
There are still some cases where the equation is obscured by the vertex. E.g. _Standard Form_ screen with a=6, b=-1, c=6:

![screenshot_769](https://user-images.githubusercontent.com/3046552/46360103-fd454680-c627-11e8-8db7-29595896baa3.png)

We should exhaustively identify these cases and decide whether to address or ignore. 

In the meantime... @ariel-phet @kathy-phet your feedback is needed with top priority.

**Comment by pixelzoom** [comment]
In case more specifics would help to understand the current equation placement algorithm...  The equation is placed on a line that is tangent to the parabola.  If you run with `?dev`, you'll see a black dot on the parabola, 0.5 units in from the edge of the graph, that indicates where the tangent point is. For example, here's a case with tangent point at (2.1,9.5):

![screenshot_770](https://user-images.githubusercontent.com/3046552/46360614-0f73b480-c629-11e8-85cc-02fdd5c65a7d.png)

When deciding whether to display the equation horizontally, we look at the vertex.  If it's <= 3.5 units from the edge of the graph, and the parabola is open towards that edge of the graph, then the equation is displayed horizontally.

**Comment by kathy-phet** [comment]
I did play with this some, and I was feeling like it did its job of labeling the equation, but that it was hard to read when it tilted because it was longer than just the mx+b used in line graphing. 

I found myself wondering: 
- would it be easier to read and follow changes if it was just horizontal near the top of the graph it was labeling, and didn't move around so much.
- could it have a slightly white see through background so it is always visible even when something else crosses behind it 
- is there a way to turn off the label (right now I don't see a way to hide/show). 

These are just thoughts though, and if you two decide to keep it tilted for consistency with line graphing that is fine too. It was just feeling hard to follow the changes or compare the two displayed equations with them able to be at different angles and orientations.

**Comment by pixelzoom** [comment]
Thanks for the feedback @kathy-phet.  Re your points:

I agree that it would be easier to read (and would have been much easier to implement!) if the parabola equations where always horizontal at the edge of the graph. 

I think a white transparent background would be helpful, especially when the equation intersects an axis or a grid line for a major tick.

Turning off the equation is the subject of https://github.com/phetsims/graphing-quadratics/issues/35.

**Comment by pixelzoom** [comment]
In [1.0.0-dev.19](https://phet-dev.colorado.edu/html/graphing-quadratics/1.0.0-dev.19/phet/graphing-quadratics_en_phet.html), I've added a translucent (opacity=0.85) white background behind equations.

**Comment by ariel-phet** [comment]
I like the white transparent background.

Since there are only 2 curves at one time maximum on most screens, I think I would be in favor of a horizontal readout that hangs near the top right or top left of the curve, depending on where the curve is on the graph. I think with some appropriate proximity to the graph, and some color coding, it will be completely clear which curve the equation is referring to even if it is just "in the vicinity" of the curve.

**Comment by pixelzoom** [comment]
Since the question of "why place the equations on the tangent?" has come up more than once on Slack... I believe the the motivation was primarily to minimize overlap, and secondarily to be consistent with other sims (Graphing Lines, Graphing Slope-Intercept).  @amanda-phet correct me if I'm wrong.

**Comment by amanda-phet** [comment]
I think the white transparent background is working well. The layering with respect to coordinates and other curves looks good to me. 

While I agree that a horizontal display might be easier to read at times, I think the reasons not to do that are significant. There is much less overlap when it is done this way than there would be otherwise, and the equations are always close in proximity to their curves. If we only ever had one curve on the graph at a time we wouldn't have explored this option. 

I think we can close this issue, but if any issues come up in interviews next week I will reopen it.
